ARIA Live Article (#193)
d-koppenhagen/d-koppenhagen.de: src/content/blog/2025-11-aria-live.md

published: true
author:
  name: 'Danny Koppenhagen'
  mail: mail@k9n.dev
created: 2025-11-05
updated: 2025-11-05
keywords:
  - accessibility
  - aria-live
  - screen reader
  - Angular
  - Vue
  - React
  - SPA
  - a11y
  - CDK
language: en
thumbnail:
  header: images/blog/aria-live/aria-live.jpg
  card: images/blog/aria-live/aria-live-small.jpg
# linked:
#   devTo: 'https://foo.bar'
---

You've built a modern single-page application with dynamic content alerts and live tickers - of course: with accessibility in mind.
Therefore, you've added `aria-live` regions so screen reader users can hear what's changing.
A success message here, a toast there.
It *should* just work.

But when you test it with a screen reader… nothing.
Silence.
Your "live" region isn't so live after all.

If that sounds familiar, you're not alone.

> Accessibility professionals and framework developers alike run into this issue across **Angular**, **Vue**, **React** and other frameworks.

Author: ```suggestion
Accessibility professionals and framework developers alike run into this issue across Angular, Vue, React and other frameworks.
```

It only reacts to **changes in the actual DOM text** inside an element that already has `aria-live` on it.
If that element doesn't exist yet, or is about to be replaced, your announcement vanishes into thin air.
In SPAs, it's common to show or hide UI elements conditionally:

```html
<!-- Angular -->
@if (showMessage) {
<div aria-live="polite">{{ message }}</div>
}

<!-- Vue -->
<div v-if="showMessage" aria-live="polite">{{ message }}</div>

<!-- React -->
{showMessage && <div aria-live="polite">{message}</div>}
```

That looks fine — but when `showMessage` changes from `false` to `true`, the framework **creates a brand new element in the DOM**.
From the screen reader's perspective, that's just *a new element appearing*, not an update in a live region it's been tracking.
And since the text `"Saved!"` is already present when the node appears, the screen reader never gets a "text change" event — so it says nothing.

So, how can we fix it? To make `aria-live` work reliably, the element:

1. Must **always exist in the DOM** (no conditional rendering), and
2. Must have **its text content changed dynamically**, not replaced by a new node.

> That's why we'll look next at two approaches:

Reviewer: ```suggestion
That's why we'll look at two approaches next:
```
Spezifische Zeitangaben stehen im Englischen meistens am Satzende

### `aria-relevant` — Controlling *What* Triggers an Announcement

The `aria-relevant` attribute refines what types of changes should be announced.
It accepts values like `additions`, `removals`, `text`, or `all`.
For most live regions, the default (`aria-relevant="additions text"`) is ideal — it announces when new content is added or existing text changes.
However, in cases where your region replaces its text frequently (e.g., progress bars, timers, or captions), you might explicitly set:

```html
<div aria-live="polite" aria-relevant="text">0% complete</div>
```

This prevents redundant announcements for removed nodes, focusing only on text changes.
You can also combine values for fine control:

```html
<div aria-live="assertive" aria-relevant="additions removals text">Error occurred</div>
```

### `aria-atomic` — Controlling *How Much* Gets Announced

The `aria-atomic` attribute determines whether the screen reader should announce only the changed part of a live region or the entire content.

- **`aria-atomic="false"`** (default)
  Only announces the specific text that changed.

>   Good for regions where you append new content (like chat messages or logs).

Author: Hab ein Beispiel zu Chat-Nachrichten ergänzt

### Vue 3

For Vue applications, I recommend using [vue-a11y/vue-announcer](https://github.com/vue-a11y/vue-announcer).

```bash
npm install @vue-a11y/announcer@next # Vue 3
# OR:
npm install @vue-a11y/announcer      # Vue 2
```

Once installed, setup the `VueAnnouncer` for your `App`.

```ts
// main.ts
import { createApp } from 'vue'
import App from './App.vue'
import router from './router'

import VueAnnouncer from '@vue-a11y/announcer'
import '@vue-a11y/announcer/dist/style.css'

createApp(App)
  .use(VueAnnouncer)
  .use(router)
  .mount('#app');
```

After that, place the component containing the global live region(s) in your main component:

```html
<!-- App.vue -->
<template>
  <VueAnnouncer class="sr-only" />
  ...
</template>
```

The last step is to use the composable `useAnnouncer` which pushes messages into the live region:

```html
<!-- SaveButton.vue -->
<template>
  <button @click="onSave">Save</button>
</template>

<script setup>
import { useAnnouncer } from '@vue-a11y/announcer'

const { polite } = useAnnouncer()

function onSave() {
  polite('Settings saved successfully.')
}
</script>
```

### React

> For React, I recommend using [@react-aria/live-announcer](https://react-spectrum.adobe.com/react-aria/useAnnouncer.html):

Reviewer: Der Link geht ins Leere und ich kann in den aktuellen Docs von React Spektrum auch kein useAnnouncer mehr finden 😮

---
title: "When Your Live Region Isn't Live: Fixing aria-live in Angular, React, and Vue"
description: 'Learn how to fix aria-live regions that fail silently in modern SPAs. Discover why screen readers miss your announcements when frameworks recreate DOM elements, understand the difference between polite and assertive announcements, and implement two reliable patterns - local and global live regions - with concrete examples for Angular, Vue, and React applications.'
published: true
author:
  name: 'Danny Koppenhagen'
  mail: mail@k9n.dev
created: 2025-11-04
updated: 2025-11-04
keywords:
  - accessibility
  - aria-live
  - screen reader
  - Angular
  - Vue
  - React
  - SPA

>   - a11y

Author: ```suggestion
  - a11y
  - CDK
```

#   devTo: 'https://foo.bar'
---

You've built a modern single-page application with dynamic content alerts and live tickers - of course: with accessibility in mind.
Therefore, you've added `aria-live` regions so screen reader users can hear what's changing.
A success message here, a toast there.
It *should* just work.

But when you test it with a screen reader… nothing.
Silence.
Your "live" region isn't so live after all.

If that sounds familiar, you're not alone.
Accessibility professionals and framework developers alike run into this issue across **Angular**, **Vue**, **React** and other frameworks.
The problem isn't your markup — it's how these frameworks manage the DOM.

Modern SPA frameworks do amazing things behind the scenes: they mount, unmount, and patch elements as state changes.
Unfortunately, screen readers don't see your reactive data; they only notice *actual DOM mutations*.
When the element holding your `aria-live` attribute is recreated or removed, assistive technologies lose track — and your updates are never announced.

In this post, we'll break down:

- Why live regions may fail in your SPAs
- The difference between **polite** and **assertive** announcements

> - What the `aria-relevant` attribute actually does

Author: ```suggestion
- What the `aria-relevant` and `aria-atomic` attributes actually doing
```

If that element doesn't exist yet, or is about to be replaced, your announcement vanishes into thin air.
In SPAs, it's common to show or hide UI elements conditionally:

```html
<!-- Angular -->
@if (showMessage) {
<div aria-live="polite">{{ message }}</div>
}

<!-- Vue -->
<div v-if="showMessage" aria-live="polite">{{ message }}</div>

<!-- React -->
{showMessage && <div aria-live="polite">{message}</div>}
```

That looks fine — but when `showMessage` changes from `false` to `true`, the framework **creates a brand new `<div>` in the DOM**.
From the screen reader's perspective, that's just *a new element appearing*, not an update in a live region it's been tracking.
And since the text `"Saved!"` is already present when the node appears, the screen reader never gets a "text change" event — so it says nothing.

So, how can we fix it? To make `aria-live` work reliably, the element:

1. Must **always exist in the DOM** (no conditional rendering), and
2. Must have **its text content changed dynamically**, not replaced by a new node.

That's why we'll look next at two approaches:

- Local live regions that stay mounted
- A global announcer that's always present

> But before that, let's clarify two critical ARIA attributes that often confuse developers: `aria-live`'s **politeness levels**, and its lesser-known partner, `aria-relevant`.

Author: ```suggestion
But before that, let's clarify three critical ARIA attributes that often confuse developers: `aria-live`'s **politeness levels**, and its lesser-known partners, `aria-relevant` and `aria-atomic`.
```

  <button @click="onSave">Save</button>
</template>

<script setup>
import { useAnnouncer } from '@vue-a11y/announcer'

const { polite } = useAnnouncer()

function onSave() {
  polite('Settings saved successfully.')
}
</script>
```

### React

For React, I recommend using [@react-aria/live-announcer](https://react-spectrum.adobe.com/react-aria/useAnnouncer.html):

```bash
npm install @react-aria/live-announcer
```

After installation, you can simply call the `announce` function which will set up the global live region if not already present and push the message into it.

```tsx
// SaveButton.tsx
import { announce } from '@react-aria/live-announcer';

function SaveButton() {
  const handleSave = () => {
    announce('Settings saved successfully.');
  };

  return <button onClick={handleSave}>Save</button>;
}
```

Alternatively, you can use the `useAnnouncer` hook for more control:

```tsx
import { useAnnouncer } from '@react-aria/live-announcer';

function SaveButton() {
  const { announce } = useAnnouncer();

  const handleSave = () => {
    // Perform save logic
    announce('Settings saved successfully.', 'polite');
  };

  return <button onClick={handleSave}>Save</button>;
}

> ```

Author: ```suggestion
```

Das gibt es so nicht.
